### PR TITLE
Feature: add ODR monitoring module

### LINKED ISSUE #13: Bootstrap Monitoring / Logging System
We need to create a new `odr_monitoring` module to handle logging of errors and information from our pipeline. This module will initially focus on early control flow issues but should be designed with future performance monitoring capabilities in mind.

Tasks:

1. Create a new `odr_monitoring` directory in the `modules` folder
2. Set up basic module structure:
   - `__init__.py`
   - `logger.py`
   - `config.py`
   - `handlers/`
   - `formatters/`
3. Implement a configurable logging system:
   - Use Python's built-in `logging` module
   - Create custom formatters for different log levels
   - Implement handlers for console and file output
4. Design an easy-to-use API for other modules to utilize the logging functionality
5. Implement basic error and info logging functions
6. Add configuration options for log levels, output destinations, and rotation policies
7. Write basic documentation for usage and configuration
9. Add basic tests for core logging functionality

Acceptance Criteria:

- [ ] `odr_monitoring` module is created and integrated into the project structure
- [ ] Logging system is implemented and configurable
- [ ] Other modules can easily use the logging functionality
- [ ] Basic error and info logging are functional
- [ ] Configuration options for logging are available and documented
- [ ] Placeholder for future performance monitoring is in place
- [ ] Tests are written and passing
- [ ] Documentation for usage and configuration is available

### PR DESCRIPTION
**Pull request by robert-cronin**
Fixes #13 

I believe I covered all the bases here, but if I've missed anything please don't hesitate to suggest improvements or fixes!

[x] odr_monitoring module is created and integrated into the project structure
[x] Logging system is implemented and configurable
[x] Other modules can easily use the logging functionality
[x] Basic error and info logging are functional
[x] Configuration options for logging are available and documented
[x] Placeholder for future performance monitoring is in place
[x] Tests are written and passing
[x] Documentation for usage and configuration is available

As for testing, I'm not sure if there has been any agreement for min coverage for new features? I am getting 87% but I can strengthen if need be! 

Thank you in advance for your review!

### CODE REVIEW

**Comment by robert-cronin** [comment]
I've just noticed that there is another logger method in api module, so I might have misinterpreted the first criterion. @fearnworks was the intention to replace all project logging with this central monitoring module as a part of this PR?

**Comment by fearnworks** [comment]
As far as coverage I don't think we are setting minimum coverage, just that we can prove out the critical path for now. 

Irt second comment : Replacing all of the logging isn't required for this feature since that could spiral out of hand. As long as  the logs are available to be used in the other components.

**Comment by fearnworks** [comment]
I get this error when running `task monitoring:test` : 

```bash
_________________________________________________________________________________________________________________________________________________________________________________________________________________________ TestODRLogger.test_logger_creation __________________________________________________________________________________________________________________________________________________________________________________________________________________________ 

self = <tests.test_logger.TestODRLogger testMethod=test_logger_creation>

    def tearDown(self):
        if os.path.exists(self.log_file):
>           os.remove(self.log_file)
E           PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'

tests\test_logger.py:46: PermissionError
=============================================================================================================================================================================================================================== short test summary info =============================================================================================================================================================================================================================== 
FAILED tests/test_logger.py::TestODRLogger::test_log_format - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
FAILED tests/test_logger.py::TestODRLogger::test_log_levels - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
FAILED tests/test_logger.py::TestODRLogger::test_logger_creation - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
================================================================================================================================================================================================================================== 3 failed in 0.17s ================================================================================================================================================================================================================================== 
task: Failed to run task "monitoring:test": exit status 1
```

Test outputs look good! Just likely need to make sure the file is released before teardown

File: modules/odr_monitoring/Taskfile.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+version: "3"
+
+tasks:
+  default:
```
Comment (fearnworks) [inline modules/odr_monitoring/Taskfile.yml:4]: This is nice. Making a mental note to propogate to the other task files

**Comment by robert-cronin** [comment]
> I get this error when running `task monitoring:test` :
> 
> ```shell
> _________________________________________________________________________________________________________________________________________________________________________________________________________________________ TestODRLogger.test_logger_creation __________________________________________________________________________________________________________________________________________________________________________________________________________________________ 
> 
> self = <tests.test_logger.TestODRLogger testMethod=test_logger_creation>
> 
>     def tearDown(self):
>         if os.path.exists(self.log_file):
> >           os.remove(self.log_file)
> E           PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
> 
> tests\test_logger.py:46: PermissionError
> =============================================================================================================================================================================================================================== short test summary info =============================================================================================================================================================================================================================== 
> FAILED tests/test_logger.py::TestODRLogger::test_log_format - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
> FAILED tests/test_logger.py::TestODRLogger::test_log_levels - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
> FAILED tests/test_logger.py::TestODRLogger::test_logger_creation - PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'logs/odr_monitoring.log'
> ================================================================================================================================================================================================================================== 3 failed in 0.17s ================================================================================================================================================================================================================================== 
> task: Failed to run task "monitoring:test": exit status 1
> ```
> 
> Test outputs look good! Just likely need to make sure the file is released before teardown

Good catch, I think you're right, we just need to release the file before teardown, I'll push a fix :+1:

File: modules/odr_monitoring/odr_monitoring/handlers/console_handler.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import logging
+import sys
+from ..config import logging_config
```
Comment (fearnworks) [inline modules/odr_monitoring/odr_monitoring/handlers/console_handler.py:3]: can we shift this to absolute module paths? i.e odr_monitoring.config , odr_monitoring.formatters.standard_formatter, etc

**Review by fearnworks — CHANGES_REQUESTED** [review]
Code looks solid. Very minor update request for the module imports

File: modules/odr_monitoring/odr_monitoring/handlers/console_handler.py (hunk 2 above)
Comment (robert-cronin) [reply modules/odr_monitoring/odr_monitoring/handlers/console_handler.py:3]: Sure thing, I'll make the change

File: modules/odr_monitoring/Taskfile.yml (hunk 1 above)
Comment (robert-cronin) [reply modules/odr_monitoring/Taskfile.yml:4]: Yeah its a cool pattern I've adopted recently, I think it just gets rid of the error upon running `tasks`

**Comment by robert-cronin** [comment]
@fearnworks latest change should fix the PR comments, don't hesitate to add more if needed, thanks!

**Comment by fearnworks** [comment]
LGTM. Tests pass. Output looks solid. Nice work! Merging

**Review by fearnworks — APPROVED** [review]

**Comment by hipsterusername** [comment]
Hey @robert-cronin - shoot us a note on discord!

**Pull request merged**